server/sonar-server/src/main/java/org/sonar/server/computation/step/PersistFileSourcesStep.java

```diff
@@ -0,0 +1,226 @@
+import org.sonar.core.source.db.FileSourceDto;
+import org.sonar.server.computation.ComputationContext;
+import org.sonar.server.computation.source.ReportIterator;
+import org.sonar.server.computation.source.StreamLine;
+import org.sonar.server.computation.source.StreamLineCoverage;
+import org.sonar.server.computation.source.StreamLineScm;
+import org.sonar.server.db.DbClient;
+import org.sonar.server.source.db.FileSourceDb;
+
+import java.io.File;
+import java.io.IOException;
+import java.security.MessageDigest;
+import java.util.HashMap;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+
+import static com.google.common.base.Charsets.UTF_8;
+import static com.google.common.collect.Lists.newArrayList;
+
+public class PersistFileSourcesStep implements ComputationStep {
+
+  private final DbClient dbClient;
+  private final System2 system2;
+
+  public PersistFileSourcesStep(DbClient dbClient, System2 system2) {
+    this.dbClient = dbClient;
+    this.system2 = system2;
+  }
+
+  @Override
+  public String[] supportedProjectQualifiers() {
+    return new String[] {Qualifiers.PROJECT};
+  }
+
+  @Override
+  public void execute(ComputationContext context) {
+    int rootComponentRef = context.getReportMetadata().getRootComponentRef();
+    // Don't use batch insert for file_sources since keeping all data in memory can produce OOM for big files
+    DbSession session = dbClient.openSession(false);
+    try {
+      final Map<String, FileSourceDto> previousFileSourcesByUuid = new HashMap<>();
+      session.select("org.sonar.core.source.db.FileSourceMapper.selectHashesForProject", context.getProject().uuid(), new ResultHandler() {
+        @Override
+        public void handleResult(ResultContext context) {
+          FileSourceDto dto = (FileSourceDto) context.getResultObject();
+          previousFileSourcesByUuid.put(dto.getFileUuid(), dto);
+        }
+      });
+
+      recursivelyProcessComponent(new FileSourcesContext(session, context, previousFileSourcesByUuid), rootComponentRef);
+    } finally {
+      MyBatis.closeQuietly(session);
+    }
+  }
+
+  private void recursivelyProcessComponent(FileSourcesContext fileSourcesContext, int componentRef) {
+    BatchReportReader reportReader = fileSourcesContext.context.getReportReader();
+    BatchReport.Component component = reportReader.readComponent(componentRef);
+    if (component.getType().equals(Constants.ComponentType.FILE)) {
+      LineIterator linesIterator = linesIterator(reportReader.readFileSource(componentRef));
+      try {
+        persistSource(fileSourcesContext, linesIterator, component, createStreamLines(reportReader, componentRef));
+      } finally {
+        linesIterator.close();
+      }
+    }
+
+    for (Integer childRef : component.getChildRefList()) {
+      recursivelyProcessComponent(fileSourcesContext, childRef);
+    }
+  }
+
+  private List<StreamLine> createStreamLines(BatchReportReader reportReader, int componentRef) {
```

Comment:
I do not understand your second question. For the first one, ok for lineReaders.